projects/bms_carrier/test/test_bps_heartbeat.c

```diff
@@ -0,0 +1,117 @@
+#include <string.h>
+
+#include "bms_events.h"
+#include "bps_heartbeat.h"
+#include "can.h"
+#include "can_unpack.h"
+#include "delay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "interrupt.h"
+#include "log.h"
+#include "ms_test_helpers.h"
+#include "soft_timer.h"
+
+#undef BPS_HB_FREQ_MS
+#define BPS_HB_FREQ_MS 50
+
+static CanStorage s_can_storage = { 0 };
+static CanSettings s_can_settings = {
+  .device_id = SYSTEM_CAN_DEVICE_BMS_CARRIER,
+  .bitrate = CAN_HW_BITRATE_500KBPS,
+  .rx_event = BMS_CAN_EVENT_RX,
+  .tx_event = BMS_CAN_EVENT_TX,
+  .fault_event = BMS_CAN_EVENT_FAULT,
+  .tx = { GPIO_PORT_A, 12 },
+  .rx = { GPIO_PORT_A, 11 },
+  .loopback = true,
+};
+
+static BpsStorage s_storage = { 0 };
+
+static uint8_t s_hb_count = 0;
+static CanAckStatus s_ack_status = NUM_CAN_ACK_STATUSES;
+static uint8_t s_bps_bitmask = 0;
+
+static StatusCode prv_hb_rx(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  s_hb_count++;
+
+  *ack_reply = s_ack_status;
+
+  CAN_UNPACK_BPS_HEARTBEAT(msg, &s_bps_bitmask);
+
+  return s_ack_status == CAN_ACK_STATUS_OK ? STATUS_CODE_OK : STATUS_CODE_TIMEOUT;
+}
+
+static uint8_t s_fault_bps_bitmask = 0;
+static bool s_fault_bps_clear = false;
+
```

Comment:
Should probz move these above `prv_hb_rx`

Reply:
moved to aggregate static declarations